connmgr.go

```diff
@@ -359,9 +367,41 @@ func (cm *BasicConnMgr) getConnsToClose() []network.Conn {
 
 	target := ncandidates - cm.cfg.lowWater
 
-	// slightly overallocate because we may have more than one conns per peer
-	selected := make([]network.Conn, 0, target+10)
+	// overallocate because we may have more than one conns per peer
+	selected := make([]network.Conn, 0, 2*target)
+	seen := make(map[network.Conn]struct{})
 
+	// first select connections that are:
+	// i) older than 10 minutes but still haven't see a stream.
+	// ii) haven't seen a new stream since 10 minutes and have NO streams open.
+	for _, inf := range candidates {
+		if target <= 0 {
+			break
+		}
+
+		// lock this to protect from concurrent modifications from connect/disconnect events
+		s := cm.segments.get(inf.id)
+		s.Lock()
+
+		for c, info := range inf.conns {
+			// connections that are older than 10 minutes but still haven't see a stream.
+			if info.lastStreamOpen.IsZero() && time.Since(info.startTime) > maxStreamOpenDuration {
+				selected = append(selected, c)
+				target--
+				seen[c] = struct{}{}
+			}
+
+			// connections that haven't seen a new stream since 10 minutes and have NO streams open.
+			if !info.lastStreamOpen.IsZero() && time.Since(info.lastStreamOpen) > maxStreamOpenDuration && info.nStreams == 0 {
```

Comment:
put the count check before the time.Since check as it is a lot cheaper and can short-circuit; you can even put it first.

Reply:
Also, cache the current time and use that instead of doing gettimeofday 1ktimes.

Reply:
good point.